src/bin/ion/commands/generate/generator.rs

```diff
@@ -229,39 +280,100 @@ impl<'a, L: Language + 'static> CodeGenerator<'a, L> {
-        let mut tera_fields = vec![];
         let mut code_gen_context = CodeGenContext::new();
-        let mut nested_anonymous_types = vec![];
+        code_gen_context.with_data_model_node(DataModelNode {
+            name: type_name.to_case(Case::UpperCamel),
+            code_gen_type: None,
+            nested_types: vec![],
+        });
+        // add the current type name into fully qualified type name path
+        if let Some(ref mut fully_qualified_type_name) = self.current_type_fully_qualified_name {
+            fully_qualified_type_name.push(type_name.to_case(Case::UpperCamel));
+        }
         let constraints = isl_type.constraints();
         for constraint in constraints {
-            self.map_constraint_to_abstract_data_type(
-                &mut nested_anonymous_types,
-                &mut tera_fields,
-                constraint,
-                &mut code_gen_context,
-            )?;
+            self.map_constraint_to_abstract_data_type(constraint, &mut code_gen_context, isl_type)?;
         }
 
         // TODO: verify the `occurs` value within a field, by default the fields are optional.
-        if let Some(abstract_data_type) = &code_gen_context.abstract_data_type {
-            // Add the nested type into parent type's tera context
-            nested_types.push(NestedType {
-                target_kind_name: type_name.to_case(Case::UpperCamel),
-                fields: tera_fields,
-                abstract_data_type: abstract_data_type.to_owned(),
-                nested_types: nested_anonymous_types,
-            });
+        // TODO: Check if abstract data type is sequence and return a fully qualified name for it accordingly.
+        if let Some(data_model_node) = &code_gen_context.data_model_node {
+            // TODO: Add the nested type into parent type's tera context
+            // TODO: Add this `data_model_node` to the data model store on the current code generation context
+            if let Some(abstract_data_type) = &data_model_node.code_gen_type {
+                let fully_qualified_type_ref = match abstract_data_type {
+                    AbstractDataType::WrappedSequence(seq) => {
+                        if seq.sequence_type.is_none() || seq.element_type.is_none() {
+                            return invalid_abstract_data_type_error(
+                                "Currently code generation does not support open ended types. \
+            Error can be due to a missing `type` or `element` constraint in the type definition.",
+                            );
+                        }
+                        L::target_type_as_sequence(
+                            abstract_data_type.fully_qualified_type_ref().ok_or(
+                                invalid_abstract_data_type_raw_error(
+                                    "Can not determine fully qualified name for the data model",
+                                ),
+                            )?,
```

Comment:
This is something that should be addressed earlier in the process. The very existence of `invalid_abstract_data_type_*` seems like an anti-pattern to me. If the data model is invalid, then it should not be possible to construct it that way in the first place.